### PR TITLE
Allow different users, and have user assets

### PR DESCRIPTION
**Pull request by XPerianer**

### CODE REVIEW

File: assets/primitives/asset_user.json [hunk 1]
```diff
@@ -0,0 +1,604 @@
...
+    145,
+    143,
+    141
+  ],
+  "spring_constants": {
+  },
+  "mounted_users": [
+    {
+      "id": 140,
+      "name": "boy",
```
Comment (birneamstiel) [inline assets/primitives/asset_user.json:583]: you use the name to load the `skp` file, right? If so, please consider using the whole filename in here...

File: src/models/model_storage.rb [hunk 2]
```diff
@@ -84,4 +80,30 @@ def bottle_specifications
...
   end
+
+  def possible_attachable_user_names
+    return @possible_names unless @possible_names.nil?
+
+    @possible_names = []
+    puts 'Loaded files for attachable users:'
+    Dir.glob(ProjectHelper.component_directory + '/attachable_users/*.skp') do |model_file_path|
+      # File scheme is '.../number-name-short_name-weight_in_grams.skp'
+      # TODO: Read in the weight
```
Comment (birneamstiel) [inline src/models/model_storage.rb:91]: please add a sample file name in this comment

Comment (birneamstiel) [reply src/models/model_storage.rb:91]: and why are `child.skp` and `boy.skp` not following this scheme? Is this part of the todo comment?

File: src/models/user_indicator_model.rb [hunk 3]
```diff
@@ -2,11 +2,10 @@
 class UserIndicatorModel
   attr_reader :definition
 
-  def initialize
+  def initialize(name: 'child')
```
Comment (birneamstiel) [inline src/models/user_indicator_model.rb:5]: Also here: I'd prefer whole filename

File: src/sketchup_objects/hub.rb [hunk 4]
```diff
@@ -87,13 +87,20 @@ def update_user_indicator
     Sketchup.active_model.start_operation('Hub: Update User Indicator', true)
     point = Geom::Point3d.new(@position)
     point.z += 1
-    model = ModelStorage.instance.models['user_indicator']
     translation = Geom::Transformation.translation(point)
-    rotation = Geom::Transformation.rotation(Geom::Point3d.new, Geom::Vector3d.new(0, 0, 1), @user_rotation)
-    @user_indicator = Sketchup.active_model.active_entities.add_instance(model.definition, translation * rotation)
+    @user_indicator = Sketchup.active_model.active_entities.add_instance(@user_indicator_definition, translation * additional_transformation *@user_transformation)
```
Comment (birneamstiel) [inline src/sketchup_objects/hub.rb:91]: missing space
```suggestion
    @user_indicator = Sketchup.active_model.active_entities.add_instance(@user_indicator_definition, translation * additional_transformation * @user_transformation)
```

File: src/sketchup_objects/hub.rb [hunk 5]
```diff
@@ -205,9 +212,11 @@ def sensor?
     @is_sensor
   end
 
-  def attach_user(force)
+  def attach_user(force, name:)
```
Comment (birneamstiel) [inline src/sketchup_objects/hub.rb:215]: I guess the colon is not necessary?

File: src/tools/place_user_tool.rb [hunk 6]
```diff
@@ -1,6 +1,9 @@
 # Places a user into the geometry i.e. someone who is injecting force into the system. This tool simulates the system
 # and opens a panel that shows information and the possibility to change parameters of the springs.
 class PlaceUserTool < Tool
+
```
Comment (birneamstiel) [inline src/tools/place_user_tool.rb:4]:
```suggestion
```

File: src/tools/place_user_tool.rb [hunk 7]
```diff
@@ -12,8 +15,19 @@ def onLButtonDown(_flags, x, y, view)
     obj = @mouse_input.snapped_object
     if !obj.nil? && obj.is_a?(Node)
       @hub = obj.hub
-      # TODO: remove this default force value here
-      @hub.is_user_attached ? @hub.remove_user : @hub.attach_user(100)
+      # TODO remove this default force value here
```
Comment (birneamstiel) [inline src/tools/place_user_tool.rb:18]: maybe remove default force value from call and put it into constant at top of class.

File: src/tools/place_user_tool.rb [hunk 8]
```diff
@@ -24,12 +38,23 @@ def onMouseMove(_flags, x, y, view)
     @mouse_input.update_positions(view, x, y)
   end
 
-  def onKeyDown(key, _repeat, _flags, _view)
+  def onKeyDown(key, _repeat, flags, _view)
```
Comment (birneamstiel) [inline src/tools/place_user_tool.rb:41]: why did you remove the underscore here, I think it's still unused, isn't it?

```suggestion
  def onKeyDown(key, _repeat, _flags, _view)
```

File: src/utility/json_export.rb [hunk 9]
```diff
@@ -53,4 +53,17 @@ def self.edges_to_hash(edges)
       }
     end
   end
+
+  def self.mounted_users_to_hash(nodes)
+    users = []
+    nodes.each do |id, node|
+      next unless node.hub.is_user_attached
```
Comment (birneamstiel) [inline src/utility/json_export.rb:60]:
```suggestion
      next unless node.hub.is_user_attached
      
```

File: src/models/model_storage.rb (hunk 2 above)
Comment (XPerianer) [reply src/models/model_storage.rb:91]: I now did the todo, so now they follow :D 
The comment was unfortunatly also just a copy paste, now I made it fit

File: src/tools/place_user_tool.rb (hunk 7 above)
Comment (XPerianer) [reply src/tools/place_user_tool.rb:18]: This is resolved due to new changes reading it in from the filename

File: src/tools/place_user_tool.rb (hunk 8 above)
Comment (XPerianer) [reply src/tools/place_user_tool.rb:41]: You are right, I played around with moving the user with ctrl + arrow keys, but that did not work, but forgot to remove the change

**Comment by XPerianer** [comment]
> 
> 
> Minor things code wise, tested the functionality: works! It's a bit slow (as discussed) but lets see later what we can do to improve that. Quick usability question: How can you change between different assets?

Right now you can't theres only one. We could make multiple buttons...
> PS: Please check the rubocop annotations.
Will do again, that really looks horrible 🐙

File: src/sketchup_objects/hub.rb (hunk 5 above)
Comment (XPerianer) [reply src/sketchup_objects/hub.rb]: Now it is ```atach_user(weight: nil, name:)``` and then it makes sense, cause you can then set the name without setting the weight...

I just wanted to make sure it is named so you can set it single.

File: src/models/user_indicator_model.rb (hunk 3 above)
Comment (XPerianer) [reply src/models/user_indicator_model.rb:5]: I agree, I think I like a refactoring where we initialize the UserIndicatorModel with the filename, and optional weight parameters, and then it either uses the default weight from the filename, or the given one.
I just did not want to have weight and filename doubled to much. (cause it seems reduntant), but if we think of the weight in the filename as the default one, everything works

File: assets/primitives/asset_user.json (hunk 1 above)
Comment (XPerianer) [reply assets/primitives/asset_user.json:583]: See below, agreed

**Comment by XPerianer** [comment]
@birneamstiel Please have another look, also if you find any usability bugs, but it seems to work resonably now (I hope). The different default weights are also implemented.

File: src/utility/json_import.rb [hunk 10]
```diff
@@ -37,6 +37,7 @@ def at_position(path, position, angle: 0, scale: 1)
       add_pods_if_near_ground(nodes)
 
       animation = json_objects['animation'].to_s
+      add_mounted_users json_objects, nodes, Geom::Transformation.rotation(Geom::Point3d.new, Geom::Vector3d.new(0,0,-1), angle)
```
Comment (birneamstiel) [inline src/utility/json_import.rb:40]: rubocop: spaces

File: src/utility/json_import.rb [hunk 11]
```diff
@@ -276,6 +282,17 @@ def add_pods(json_objects, nodes)
       end
     end
 
+    def add_mounted_users(json_objects, nodes, rotation_around_center)
+      return if json_objects['mounted_users'].nil?
+      json_objects['mounted_users'].each do |user_json|
+        node = nodes[user_json['id']]
+        node.hub.attach_user weight: user_json['weight'], filename: user_json['filename']
```
Comment (birneamstiel) [inline src/utility/json_import.rb:289]: This will fail on the server side because the weight is set as string (but should be int)
```suggestion
        node.hub.attach_user weight: user_json['weight'].to_i, filename: user_json['filename']
```

**Comment by birneamstiel** [comment]
Seems better now but there are still some issues... Did you check if it's working? For me it's hard to check because the simulation is super slow...

File: src/tools/import_tool.rb [hunk 12]
```diff
@@ -67,7 +68,12 @@ def onLButtonDown(_flags, x, y, view)
     snapped_object = @mouse_input.snapped_object
     import_from_json(@path, snapped_object, @mouse_input.position)
     @ui.animation_pane.sync_hidden_status(Graph.instance.actuator_groups)
-    @ui.spring_pane.update_springs
+
+    if @update_springs
+      @ui.spring_pane.update_mounted_users
+      @ui.spring_pane.compile
+      @ui.spring_pane.update_springs
+    end
```
Comment (birneamstiel) [inline src/tools/import_tool.rb:76]:
```suggestion
    if @update_springs
      @ui.spring_pane.compile
      @ui.spring_pane.update_mounted_users
      @ui.spring_pane.update_springs
    end
```

**Comment by XPerianer** [comment]
@robertkovax  Everything is relativly slow right now, but I think it works okay. Maybe you wanna have a look at it?

File: src/tools/place_user_tool.rb [hunk 13]
```diff
@@ -12,8 +14,19 @@ def onLButtonDown(_flags, x, y, view)
...
+          @hub.remove_user
+        else
+          @hub.attach_user(filename: possible_filenames[current_index + 1])
+        end
+      else
+        @hub.attach_user(filename: possible_filenames[0])
+      end
       # TODO: at some point springe pane should compile automatically when geometry changes
       @ui.spring_pane.compile
       @ui.spring_pane.update_mounted_users
```
[the remark this answers is not in the document]
Comment (XPerianer) [reply src/tools/place_user_tool.rb]: Update to that: Compiling is needed cause of the change of user_weight, so we need to do that. (At least update the user)
